File: fsharp/interest-is-interesting/InterestIsInteresting.fs
module InterestIsInteresting

let interestRate = function
    | balance when balance < 0m -> 3.213f
    | balance when balance < 1000m -> 0.5f
    | balance when balance < 5000m -> 1.621f
    | _ -> 2.475f

let interest (balance: decimal): decimal =
   balance * decimal (interestRate balance) / 100m

let annualBalanceUpdate(balance: decimal): decimal =
   balance + interest balance

let amountToDonate(balance: decimal) (taxFreePercentage: float): int =
   if balance < 0m then 0
   else balance * decimal taxFreePercentage / 50m |> int

File: go/interest-is-interesting/interest_is_interesting.go
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) (rate float32) {
	switch {
	case balance < 0:
		rate = 3.213
	case balance < 1000:
		rate = 0.5
	case balance < 5000:
		rate = 1.621
	default:
		rate = 2.475
	}
	return
}

// Interest calculates the interest for the provided balance.
func Interest(balance float64) float64 {
	return balance * float64(InterestRate(balance)) / 100
}

// AnnualBalanceUpdate calculates the annual balance update, taking into account the interest rate.
func AnnualBalanceUpdate(balance float64) float64 {
	return balance + Interest(balance)
}

// YearsBeforeDesiredBalance calculates the minimum number of years required to reach the desired balance.
func YearsBeforeDesiredBalance(balance, targetBalance float64) (years int) {
	for balance < targetBalance {
		years++
		balance = AnnualBalanceUpdate(balance)
	}
	return years
}